File: net/protocol/src/wanProfiles.ts
// Named WAN profiles — the shared, transport-agnostic transport-condition presets
// for #159. Each profile bundles a base one-way delay, a jitter config, and a loss
// rate; the web `TransportShim` and the Bevy `Conditioner` both translate a profile
// into their own per-direction link config, so "4g-mobile" means the same impairment
// on both stacks.
//
// PARITY / HONESTY notes baked in:
// - Injection is SYMMETRIC up/down (one profile → both directions) — matches the
//   existing `sym(...)` shim points; asymmetric routes are out of scope for #159.
// - REORDER has no field on purpose. Packet reordering EMERGES from jitter (a later
//   sample can get a smaller delay and overtake an earlier one); netem itself says
//   reordering requires variable delay. On the web side this emergent reorder is an
//   APPROXIMATION (Colyseus is a reliable ordered channel — real web transport would
//   suppress it via HOL blocking); on the Bevy side (UDP) it is faithful. Documented,
//   not faked (Core Value #1). See `net/CLAUDE.md` / COMPARISON §8.
// - These are localhost-injected presets, NOT a WAN benchmark. Grounding sources are
//   in each profile's `notes`.

import type { JitterConfig } from './jitter.js';

/** Profile ids. `clean` is the zero-injection control. */
export const WAN_PROFILE_IDS = [
  'clean',
  'good-wifi',
  '4g-mobile',
  'transcontinental',
] as const;

export type WanProfileId = (typeof WAN_PROFILE_IDS)[number];

/** A named transport-condition preset (applied symmetrically up + down). */
export interface WanProfile {
  /** Stable id (also the scenario-stage label). */
  id: WanProfileId;
  /** Human-readable label. */
  label: string;
  /** Base one-way delay, ms (RTT ≈ 2× this). */
  oneWayDelayMs: number;
  /** Jitter added on top of the base delay, per delivery. */
  jitter: JitterConfig;
  /** Packet loss, percent in [0, 100]. */
  lossPct: number;
  /** Grounding sources + honest-feel note for interpreting this profile. */
  notes: string;
}

/**
 * The profile registry. Numbers are grounded in the #159 research memo's cited
 * sources (good-/poor-network thresholds, 5G/LTE latency studies, transatlantic
 * RTT + tier-1 SLAs) — directional presets, not a measured WAN.
 */
export const WAN_PROFILES: Record<WanProfileId, WanProfile> = {
  clean: {
    id: 'clean',
    label: 'Clean (localhost control)',
    oneWayDelayMs: 0,
    jitter: { sigmaMs: 0, distribution: 'none', correlation: 0 },
    lossPct: 0,
    notes:
      'Zero-injection control — the existing localhost baseline. Isolates the ' +
      'profile sweep from the machine/transport floor.',
  },
  'good-wifi': {
    id: 'good-wifi',
    label: 'Good Wi-Fi / LAN',
    oneWayDelayMs: 10,
    jitter: { sigmaMs: 3, distribution: 'normal', correlation: 0.25 },
    lossPct: 0.1,
    notes:
      '~20 ms RTT, small symmetric ±3 ms jitter (normal), 0.1% loss. LAN/good-Wi-Fi ' +
      'is low and symmetric, so a normal distribution fits. Grounding: UC good/poor ' +
      'thresholds (Tom Talks UC, OnSIP).',
  },
  '4g-mobile': {
    id: '4g-mobile',
    label: '4G / LTE mobile',
    oneWayDelayMs: 25,
    jitter: { sigmaMs: 10, distribution: 'pareto', correlation: 0.25 },
    lossPct: 1.0,
    notes:
      '~50 ms RTT, ±10 ms long-tailed jitter (pareto — cellular scheduling spikes), ' +
      '1% loss. The long tail models occasional radio-scheduling latency bursts. ' +
      'Grounding: R&S 5G/LTE latency analysis.',
  },
  transcontinental: {
    id: 'transcontinental',
    label: 'Transcontinental / transatlantic',
    oneWayDelayMs: 80,
    jitter: { sigmaMs: 20, distribution: 'paretonormal', correlation: 0.5 },
    lossPct: 0.5,
    notes:
      '~160 ms RTT, ±20 ms jitter (paretonormal — stable core + path-change long ' +
      'tail), 0.5% loss, higher correlation (long-haul wander). Grounding: ' +
      'transatlantic RTT (Broadcast Bridge) + tier-1 SLA (NTT-GIN).',
  },
};

/** All profiles in registry order (clean first). */
export function allWanProfiles(): WanProfile[] {
  return WAN_PROFILE_IDS.map((id) => WAN_PROFILES[id]);
}

/** Look up a profile by id, or `undefined` if unknown. */
export function wanProfile(id: string): WanProfile | undefined {
  return (WAN_PROFILES as Record<string, WanProfile>)[id];
}

File: net/bevy/src/wan_profiles.rs
//! Named WAN profiles — the Bevy mirror of `net/protocol/src/wanProfiles.ts`
//! (#159). Same ids and same numbers, so "4g-mobile" means the same impairment as
//! the web stack. Kept in lockstep with the TS source (like `rng.rs` mirrors
//! `rng.ts`); the values are grounded in the #159 research memo's cited sources.
//!
//! Injection is SYMMETRIC up/down. Reorder has no field — it EMERGES from jitter
//! (faithful here: renet is UDP; only APPROXIMATE on web). These are
//! localhost-injected presets, NOT a WAN benchmark.

use std::path::Path;

use serde::Serialize;

use crate::jitter::{JitterConfig, JitterDistribution};

/// A named transport-condition preset (applied symmetrically up + down).
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct WanProfile {
    /// Stable id (also the scenario-stage label) — matches the TS `WanProfileId`.
    pub id: &'static str,
    /// Base one-way delay, ms (RTT ≈ 2× this).
    pub one_way_delay_ms: f64,
    /// Jitter added on top of the base delay, per delivery.
    pub jitter: JitterConfig,
    /// Packet loss, percent in `[0, 100]`.
    pub loss_pct: f64,
}

/// All profiles, clean first (matches the TS registry order).
pub fn all_wan_profiles() -> [WanProfile; 4] {
    [
        // clean — zero-injection control.
        WanProfile {
            id: "clean",
            one_way_delay_ms: 0.0,
            jitter: JitterConfig::NONE,
            loss_pct: 0.0,
        },
        // good-wifi — ~20ms RTT, small symmetric ±3ms normal jitter, 0.1% loss.
        WanProfile {
            id: "good-wifi",
            one_way_delay_ms: 10.0,
            jitter: JitterConfig {
                sigma_ms: 3.0,
                distribution: JitterDistribution::Normal,
                correlation: 0.25,
            },
            loss_pct: 0.1,
        },
        // 4g-mobile — ~50ms RTT, ±10ms long-tailed pareto jitter, 1% loss.
        WanProfile {
            id: "4g-mobile",
            one_way_delay_ms: 25.0,
            jitter: JitterConfig {
                sigma_ms: 10.0,
                distribution: JitterDistribution::Pareto,
                correlation: 0.25,
            },
            loss_pct: 1.0,
        },
        // transcontinental — ~160ms RTT, ±20ms paretonormal jitter, 0.5% loss.
        WanProfile {
            id: "transcontinental",
            one_way_delay_ms: 80.0,
            jitter: JitterConfig {
                sigma_ms: 20.0,
                distribution: JitterDistribution::ParetoNormal,
                correlation: 0.5,
            },
            loss_pct: 0.5,
        },
    ]
}

/// Reorder-fidelity note for the Bevy stack — faithful (UDP), mirrors the TS
/// `REORDER_NOTE.bevy`.
pub const REORDER_NOTE_BEVY: &str =
    "Reorder emerges from jitter; faithful — renet runs over UDP (out-of-order \
     delivery is physical).";

/// One manifest entry — the profile knobs the thin `MetricsSample` does NOT carry.
/// camelCase to match the web `scenario-manifest.json` (`net-protocol`) so either
/// stack's sidecar reads with the same schema.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct ManifestEntry {
    scenario: String,
    profile: &'static str,
    one_way_delay_ms: f64,
    jitter_sigma_ms: f64,
    jitter_distribution: JitterDistribution,
    jitter_correlation: f64,
    loss_pct: f64,
    reorder_note: &'static str,
}

/// The sidecar file shape (mirrors the TS `ScenarioManifest`).
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct ScenarioManifest {
    seed: u32,
    engine: &'static str,
    note: String,
    entries: Vec<ManifestEntry>,
}

/// Write the `scenario-manifest.json` sidecar for the WAN sweep — the
/// jitter/distribution/correlation the thin `MetricsSample` deliberately omits.
/// Readers LEFT JOIN onto metrics.jsonl on scenario + injectedDelay* + lossPct.
pub fn write_wan_manifest(path: &Path, scenario: &str, seed: u32) -> std::io::Result<()> {
    let entries = all_wan_profiles()
        .iter()
        .map(|p| ManifestEntry {
            scenario: scenario.to_string(),
            profile: p.id,
            one_way_delay_ms: p.one_way_delay_ms,
            jitter_sigma_ms: p.jitter.sigma_ms,
            jitter_distribution: p.jitter.distribution,
            jitter_correlation: p.jitter.correlation,
            loss_pct: p.loss_pct,
            reorder_note: REORDER_NOTE_BEVY,
        })
        .collect();
    let manifest = ScenarioManifest {
        seed,
        engine: "bevy",
        note: "WAN-profile sweep sidecar. Join onto metrics.jsonl on scenario + injectedDelay* \
               + lossPct. Jitter is reproducible from the seed (shared sampler)."
            .to_string(),
        entries,
    };
    let json = serde_json::to_string_pretty(&manifest).expect("serialize manifest");
    std::fs::write(path, json + "\n")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn clean_is_zero_injection_and_first() {
        let all = all_wan_profiles();
        assert_eq!(all[0].id, "clean");
        assert_eq!(all[0].one_way_delay_ms, 0.0);
        assert_eq!(all[0].loss_pct, 0.0);
        assert_eq!(all[0].jitter.distribution, JitterDistribution::None);
    }

    #[test]
    fn impaired_profiles_increase_delay_monotonically() {
        let all = all_wan_profiles();
        let delays: Vec<f64> = all.iter().map(|p| p.one_way_delay_ms).collect();
        let mut sorted = delays.clone();
        sorted.sort_by(f64::total_cmp);
        assert_eq!(delays, sorted, "profiles must be ordered by ascending delay");
        for p in all.iter().filter(|p| p.id != "clean") {
            assert!(p.one_way_delay_ms > 0.0);
            assert!(p.jitter.sigma_ms > 0.0);
            assert_ne!(p.jitter.distribution, JitterDistribution::None);
            assert!((0.0..1.0).contains(&p.jitter.correlation));
            assert!((0.0..100.0).contains(&p.loss_pct));
        }
    }
}
